### ISSUE TITLE
Daily Notes not showing up in Chronological Order

### ISSUE THREAD

**Issue by itsrainingmani** [athensresearch/athens#235]
**Problem**

After creating a certain number of Daily Notes, they no longer show up in chronological order.

**Expected Behavior**

Daily Notes should always be in chronological order irrespective of how many daily notes have been created

**To Reproduce**

* Go to Daily Notes page
* Create atleast 10 new daily notes by scrolling down 

**Screenshots**

https://www.loom.com/share/7fa0eeb7148d4da68e354d2e254e002b
<img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/7fa0eeb7148d4da68e354d2e254e002b-with-play.gif"> </a>

<img width="1030" alt="Screen Shot 2020-07-08 at 3 52 51 PM" src="https://user-images.githubusercontent.com/10191300/87073489-1c9cda80-c1e3-11ea-9e6c-deeeabe1a38c.png">

**Comment by itsrainingmani** [comment]
@tangjeff0 Can you assign this to me (again) 😅 ?

**Comment by tangjeff0** [comment]
You can't assign yourself? That's why I added you all to `/devs` team lol.

Also, did you get to read my comment about the "Jump to Timeline" on the original issue?

**Comment by itsrainingmani** [comment]
I wasn't able to assign myself. Not sure why. 

Unfortunately I didn't get to read it. Do you remember what you wrote?

**Comment by tangjeff0** [comment]
Using ascending dates to debug was a good move. I feel that `is-timeline-page` is hacky, but it's based on Roam's US date format and doesn't seem to be the problem here, because the re-frame vector is correct.

You should know that I would eventually like to have a "Jump to Timeline" feature, mentioned in #66. This would allow a user to jump to any date in the past and scroll up or down to go to neighboring dates. Right now in Roam, you can only scroll down starting from today.

Basically, if you are on a date page, you would see a little button like `< Timeline` near the title to take you to a scrollable date page. This may inform your solution here (you would have to append OR prepend the vector).

Also, the vector should be cleared when you navigate away from daily pages. Don't think that's the problem, just something to keep in mind.

**Comment by itsrainingmani** [comment]
Awesome! Thanks Jeff.

**Comment by itsrainingmani** [comment]
The issue is beyond just daily notes. Even using Athena to create 9 pages in a row messes up the internal order of the pages.
<img width="1279" alt="Screen Shot 2020-07-09 at 8 26 52 PM" src="https://user-images.githubusercontent.com/10191300/87106331-d2d3e480-c222-11ea-8bd5-73220133df92.png">

**Comment by nthd3gr33** [comment]
Documenting some things from convo with @itsrainingmani:

> if you fire up a clojure repl and you make a map and conj stuff into it, you'll notice that first ~8 elements have insertion order preserved as well. and the 9th insertion re-arranges the order

```clojure
;; small maps stay as 
(def mapA {:a 1 :b 2})
mapA   ; => {:a 1, :b 2}

;; somewhere around 9 elements leads to rearranging
(def mapB {:a 1 :b 2 :c 3 :d 4 :e 5 :f 6 :g 7 :h 8 :i 9})
mapB   ; => {:e 5, :g 7, :c 3, :h 8, :b 2, :d 4, :f 6, :i 9, :a 1}
```

(see this [clojuredocs.org link](https://clojuredocs.org/clojure.core/array-map#example-57392e25e4b071da7d6cfd0c) for reference)

My initial thought was to attempt to override this behavior somehow (naive thinking), but @itsrainingmani informed me that > datascript doesn't use a plain ol clojure map internally. it uses a custom implementation of a B+ Tree Structure

See this excerpt from a relevant [post about Datascript internals](https://tonsky.me/blog/datascript-internals/) from Tonsky's blog.   


![Screen Shot 2020-07-10 at 2 00 12 PM](https://user-images.githubusercontent.com/37888959/87118216-b9c84380-c2b5-11ea-9b98-5e2ace382e27.png)

We are looking into B+ tree structures starting with the [wiki](https://en.wikipedia.org/wiki/B%2B_tree). Will update as we make more progress.

**Comment by itsrainingmani** [comment]
I think the issue here is in the order that DataScript returns the results of a query rather than an issue with `daily-notes` since the `all_pages` table order also seems to be altered after 8 or so page creations.

The number 8 is interesting because the insertion order seems to be "preserved" until a 9th element is added which then changes the order. However, the DataScript indexes (eavt, avet, aevt) don't seem to change.

Internally DataScript uses a Custom B+ Tree Structure to hold the data. DataScript also maintains multiple indices to hold the datoms. The 3 indices are EAVT, AVET, AEVT.

It seems like there is no order guarantee for DataScript queries as far as I can see.

Now why does DataScript preserve order for some number of items? My hypothesis is that this is a function of the B+ Tree structure. 

<img width="1110" alt="datascript-order" src="https://user-images.githubusercontent.com/10191300/87128594-83f37280-c255-11ea-9985-11064059fd23.png">

After a certain number of elements, (in this case 8), the tree rebalances it's buckets.

<img width="1110" alt="datascript-unbalance" src="https://user-images.githubusercontent.com/10191300/87128606-881f9000-c255-11ea-9ec0-b1230d47d5b5.png">

Solutions - 

* Sort the result of the query. The tradeoff is that for large result sets, this could be expensive.
* Use the `datoms` function to query the DataScript indices directly for ordered data. https://cljdoc.org/d/datascript/datascript/0.18.13/api/datascript.core#datoms

**Comment by tangjeff0** [comment]
#20

**Comment by sid597** [comment]
Went back ~20 notes, everything is correct and can't reproduce it.
